File: src/molecular_analysis_dashboard/presentation/api/schemas/tasks.py
"""
Task API schemas for molecular analysis dashboard.

This module defines Pydantic schemas for the task management API endpoints,
providing type-safe request and response models that match the frontend interface.

Responsibilities:
- Define TaskTemplate schema matching frontend TypeScript interfaces
- Provide validation and serialization for API endpoints
- Support field aliasing for camelCase/snake_case conversion
- Define enum types for task categories, complexity, and resource requirements

Dependencies:
- pydantic: For schema validation and serialization
- typing: For type annotations
- enum: For enumerated values

Assumptions:
- Frontend uses camelCase naming convention
- Backend uses snake_case naming convention
- All API responses follow consistent structure with error handling
"""

from enum import Enum
from typing import Any, Dict, List, Optional, Union

from pydantic import BaseModel, Field


class TaskComplexity(str, Enum):
    """Task complexity levels."""

    BEGINNER = "beginner"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"


class TaskCategory(str, Enum):
    """Task categories."""

    AUTODOCK_VINA = "autodock_vina"
    AUTODOCK4 = "autodock4"
    SCHRODINGER = "schrodinger"
    CUSTOM = "custom"


class ResourceRequirement(str, Enum):
    """Resource requirement levels."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class TaskParameter(BaseModel):
    """Task parameter definition."""

    name: str = Field(..., description="Parameter name")
    type: str = Field(..., description="Parameter type (string, number, boolean, file, select)")
    required: bool = Field(..., description="Whether parameter is required")
    default: Optional[Union[str, int, float, bool]] = Field(None, description="Default value")
    description: str = Field(..., description="Parameter description")
    options: Optional[List[str]] = Field(None, description="Available options for select type")


class TaskTemplate(BaseModel):
    """Task template response schema matching frontend interface."""

    id: str = Field(..., description="Task identifier")
    name: str = Field(..., description="Task display name")
    description: str = Field(..., description="Task description")
    category: TaskCategory = Field(..., description="Task category")
    version: str = Field(..., description="Task version")
    complexity: TaskComplexity = Field(..., description="Task complexity level")
    estimated_runtime: str = Field(
        ..., alias="estimatedRuntime", description="Estimated execution time"
    )
    cpu_requirement: ResourceRequirement = Field(
        ..., alias="cpuRequirement", description="CPU requirement"
    )
    memory_requirement: ResourceRequirement = Field(
        ..., alias="memoryRequirement", description="Memory requirement"
    )
    required_files: List[str] = Field(
        ..., alias="requiredFiles", description="Required input files"
    )
    parameters: List[TaskParameter] = Field(..., description="Task parameters")
    compatibility: List[str] = Field(..., description="Supported platforms")
    tags: List[str] = Field(..., description="Task tags")
    documentation: Optional[str] = Field(None, description="Task documentation")
    examples: Optional[List[str]] = Field(None, description="Usage examples")
    is_built_in: bool = Field(..., alias="isBuiltIn", description="Whether task is built-in")

    class Config:
        """Pydantic configuration."""

        validate_by_name = True
        extra = "forbid"


class TaskListResponse(BaseModel):
    """Response schema for task list endpoint."""

    tasks: List[TaskTemplate] = Field(..., description="List of available tasks")
    total_count: int = Field(..., description="Total number of tasks")
    organization_id: str = Field(..., description="Organization ID")

    class Config:
        """Pydantic configuration."""

        extra = "forbid"


class TaskDetailResponse(BaseModel):
    """Response schema for task detail endpoint."""

    task: TaskTemplate = Field(..., description="Task details")
    api_specification: Dict[str, Any] = Field(
        ..., description="OpenAPI specification for task execution"
    )
    service_configuration: Dict[str, Any] = Field(
        ..., description="Service deployment configuration"
    )

    class Config:
        """Pydantic configuration."""

        extra = "forbid"


class ErrorResponse(BaseModel):
    """Error response schema."""

    error: str = Field(..., description="Error type")
    message: str = Field(..., description="Error message")
    code: int = Field(..., description="Error code")
    details: Optional[Dict[str, Any]] = Field(None, description="Additional error details")

    class Config:
        """Pydantic configuration."""

        extra = "forbid"

File: frontend/src/types/tasks.ts
/**
 * Task API Type Definitions
 *
 * TypeScript interfaces that match the backend API schemas exactly,
 * ensuring type safety for all task-related API communications.
 */

// Core task types matching backend schemas
export interface TaskTemplate {
  id: string;
  name: string;
  description: string;
  category: string;
  subcategory?: string;
  version: string;
  engine: string;
  status: string;
  tags: string[];
  parameters: TaskParameter[];
  outputs?: TaskOutput[];
  execution_time_estimate?: number;
  resource_requirements?: ResourceRequirements;
  created_at: string;
  updated_at: string;
}

export interface TaskParameter {
  name: string;
  type: 'string' | 'number' | 'integer' | 'boolean' | 'file' | 'select';
  required: boolean;
  default?: any;
  description: string;
  options?: string[];
  validation?: {
    min?: number;
    max?: number;
    pattern?: string;
    file_types?: string[];
    max_size_mb?: number;
    allowed_values?: string[];
  };
}

export interface TaskOutput {
  name: string;
  type: string;
  description: string;
  file_pattern?: string;
}

export interface ResourceRequirements {
  cpu_cores: number;
  memory_gb: number;
  disk_gb: number;
}

// API response types
export interface TaskListResponse {
  tasks: TaskTemplate[];
  total_count: number;
  organization_id: string;
}

export interface TaskDetailResponse {
  task: TaskTemplate;
  api_specification: Record<string, any>;
  service_configuration: Record<string, any>;
}

export interface ErrorResponse {
  error: string;
  message: string;
  code: number;
  details?: Record<string, any>;
}

// API request parameters
export interface TaskListParams {
  org_id?: string;
  category?: string;
  engine?: string;
  status?: string;
  search?: string;
  offset?: number;
  limit?: number;
}

export interface TaskDetailParams {
  task_id: string;
  org_id?: string;
}

export interface TaskCategoriesParams {
  org_id?: string;
}

// Service configuration types
export interface TaskServiceConfig {
  baseUrl: string;
  timeout: number;
  retries: number;
  fallbackEnabled: boolean;
  organizationId?: string;
}

// Feature flag types
export interface FeatureFlags {
  useApiTasks: boolean;
  enableTaskCache: boolean;
  debugMode: boolean;
}

// Error types
export class TaskApiError extends Error {
  constructor(
    public override message: string,
    public status?: number,
    public code?: string,
    public details?: any,
  ) {
    super(message);
    this.name = 'TaskApiError';
  }
}

export class TaskServiceError extends Error {
  constructor(
    public override message: string,
    public override cause?: Error,
  ) {
    super(message);
    this.name = 'TaskServiceError';
  }
}

// Service response types
export interface ServiceResponse<T> {
  data: T;
  source: 'api' | 'fallback' | 'cache';
  timestamp: number;
  error?: string;
}

// Cache types
export interface CacheEntry<T> {
  data: T;
  timestamp: number;
  expires: number;
}

export interface TaskCache {
  tasks: CacheEntry<TaskTemplate[]> | null;
  categories: CacheEntry<string[]> | null;
  details: Map<string, CacheEntry<TaskTemplate>>;
}

// UI state types
export interface TaskLoadingState {
  isLoading: boolean;
  error: string | null;
  source: 'api' | 'fallback' | 'cache' | null;
}

export interface TaskFilters {
  search: string;
  category: string;
  complexity: string;
  tags: string[];
}

// Constants
export const TASK_CATEGORIES = ['autodock_vina', 'autodock4', 'schrodinger', 'custom'] as const;

export const TASK_COMPLEXITIES = ['beginner', 'intermediate', 'advanced'] as const;

export const RESOURCE_REQUIREMENTS = ['low', 'medium', 'high'] as const;

export type TaskCategory = (typeof TASK_CATEGORIES)[number];
export type TaskComplexity = (typeof TASK_COMPLEXITIES)[number];
export type ResourceRequirement = (typeof RESOURCE_REQUIREMENTS)[number];
